### PR TITLE
Feature/create dataset

### PR DESCRIPTION
**Pull request by CarlHembrough**
### What

* Add /page POST and DELETE endpoints.
* Add page type specific hooks, allowing additional functionality to be added when creating / deleting specific page types.
* Add hooks for the creation / deletion of dataset landing pages, which call the dataset API to create / delete the dataset.
* migrated existing code to use the dp-dataset-api-java-client and removed the existing API client from Zebedee

### How to review

Review changes

### Who can review

Anyone

### CODE REVIEW

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        this.pageCreationHook = pageCreationHook;
+        this.pageDeletionHook = pageDeletionHook;
+    }
+
+    /**
+     * create a new page from the endpoint <code>/Content/[CollectionName]/?uri=[uri]</code>
+     *
+     * @param request  This should contain a X-Florence-Token header for the current session
+     * @param response Returns true or false according to whether the URI was deleted.
+     * @return
```
Comment (Matt-Guest) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: This method does not return anything

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+                    .log();
+        }
+    }
+
+    /**
+     * Deletes page content from the endpoint <code>/Content/[CollectionName]/?uri=[uri]</code>
+     *
+     * @param request  This should contain a X-Florence-Token header for the current session
+     * @param response Returns true or false according to whether the URI was deleted.
+     * @return
```
Comment (Matt-Guest) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: @return is not needed here

File: zebedee-cms/pom.xml [hunk 3]
```diff
@@ -168,6 +168,12 @@
             <version>4.3.8.RELEASE</version>
         </dependency>
 
+        <dependency>
+            <groupId>com.github.ONSdigital</groupId>
+            <artifactId>dp-dataset-api-java-client</artifactId>
+            <version>feature~initial-implementation-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (CarlHembrough) [reply zebedee-cms/pom.xml]: Yes, its dependent on the api client branch being merged 👍

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+@Api
+public class Page {
+
+    private ZebedeeCmsService zebedeeCmsService;
+    private PageUpdateHook pageCreationHook;
+    private PageUpdateHook pageDeletionHook;
+
+    static final String zebedeeFileSuffix = "/data.json";
+    static final URI websiteURI = URI.create(Configuration.getBabbageUrl());
```
[the remark this answers is not in the document]
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: website URI was redundant and removed. The zebedeeFileSuffix is also used in tests hence the package level. Do you think its better to make it private and have the test declare it itself?

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+import java.net.URISyntaxException;
+import java.util.HashMap;
+import java.util.Map;
+
+@Api
+public class Page {
+
+    private ZebedeeCmsService zebedeeCmsService;
+    private PageUpdateHook pageCreationHook;
+    private PageUpdateHook pageDeletionHook;
```
[the remark this answers is not in the document]
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java (hunk 1 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: Removed

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java (hunk 2 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: Removed

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+     * @throws ConflictException     If the URI is being edited in another collection
+     */
+    @POST
+    public void createPage(HttpServletRequest request, HttpServletResponse response) throws IOException, ZebedeeException, FileUploadException {
+
+        // We have to get the request InputStream before reading any request parameters
+        // otherwise the call to get a request parameter will actually consume the body:
+        try (InputStream requestBody = request.getInputStream()) {
+
+            Session session = zebedeeCmsService.getSession(request);
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: Can you move this into its own try catch? if there is an error log something like "error getting session", set the response status and rethrow the exception?

Tedious I know but its clear what gone on then.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+            com.github.onsdigital.zebedee.model.Collection collection = zebedeeCmsService.getCollection(request);
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: same with the collection

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+            String uri = request.getParameter("uri");
+            uri = trimZebedeeFileSuffix(uri);
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: can you check if uri is null / empty and throw the appropriate exception/ logging etc

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageCreationHook.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        taxonomyLink.setHref(uri);
+
+        DatasetLinks datasetLinks = new DatasetLinks();
+        datasetLinks.setTaxonomy(taxonomyLink);
+
+        dataset.setLinks(datasetLinks);
+
+        try {
+            datasetAPIClient.createDataset(page.getapiDatasetId(), dataset);
+        } catch (DatasetAPIException e) {
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageCreationHook.java]: add a log message and rethrow

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageDeletionHook.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+     * @param uri - the URI of the updated page.
+     * @throws IOException
+     */
+    @Override
+    public void onPageUpdated(ApiDatasetLandingPage page, String uri) throws IOException {
+
+        try {
+            datasetAPIClient.deleteDataset(page.getapiDatasetId());
+        } catch (DatasetAPIException e) {
+            throw new RuntimeException(e);
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageDeletionHook.java]: add log message

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/PageTypeUpdateHook.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * Allows different page types to be attached to different page update hooks.
+ */
+public class PageTypeUpdateHook implements PageUpdateHook<Page> {
+
+    Map<PageType, PageUpdateHook> pageUpdateHooks;
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/PageTypeUpdateHook.java]: unless there is a need can this be made private?

File: zebedee-cms/src/test/java/com/github/onsdigital/zebedee/api/PageTest.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+import java.io.IOException;
+import java.nio.file.Paths;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class PageTest extends ZebedeeAPIBaseTestCase {
```
Comment (daiLlew) [inline zebedee-cms/src/test/java/com/github/onsdigital/zebedee/api/PageTest.java]: can you add test cases to cover session/collection not found exceptions & uri empty null?

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+
+            byte[] bytes = IOUtils.toByteArray(requestBody);
+            com.github.onsdigital.zebedee.content.page.base.Page page;
+
+            try (ByteArrayInputStream pageInputStream = new ByteArrayInputStream(bytes)) {
+                page = ContentUtil.deserialiseContent(pageInputStream);
+            } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+            try (ByteArrayInputStream inputStream = new ByteArrayInputStream(bytes)) {
+                zebedeeCmsService.getZebedee().getCollections().createContent(
+                        collection,
+                        uri + zebedeeFileSuffix,
+                        session,
+                        request,
+                        inputStream,
+                        CollectionEventType.COLLECTION_PAGE_SAVED,
+                        false);
+            }
```
[the remark this answers is not in the document]
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: I thought Zebedee had an error handler but that doesn't appear to be the case. I have added handling as suggested as its more explicit.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+     * @throws BadRequestException   If the request cannot be completed because of a problem with request parameters
+     * @throws NotFoundException     If the requested URI does not exist in the collection.
+     * @throws UnauthorizedException If the user does not have publisher permission.
+     * @throws ConflictException     If the URI is being edited in another collection
+     */
+    @DELETE
+    public boolean deletePage(HttpServletRequest request, HttpServletResponse response) throws IOException,
+            ZebedeeException {
+
+        Session session = zebedeeCmsService.getSession(request);
```
[the remark this answers is not in the document]
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+
+        Collection collection = zebedeeCmsService.getCollection(request);
```
[the remark this answers is not in the document]
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java (hunk 6 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: Fair point. I have caught exceptions locally throughout and added contextual messages. I have not thrown any exceptions out of the handler functions as the response status is already set.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java (hunk 7 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java (hunk 8 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageCreationHook.java (hunk 9 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageCreationHook.java]: Added log message. Did not rethrow as the response code is set locally.

Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageCreationHook.java]: done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/PageTypeUpdateHook.java (hunk 11 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/PageTypeUpdateHook.java]: made private and final 👍

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageDeletionHook.java (hunk 10 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageDeletionHook.java]: done

File: zebedee-cms/src/test/java/com/github/onsdigital/zebedee/api/PageTest.java (hunk 12 above)
Comment (CarlHembrough) [reply zebedee-cms/src/test/java/com/github/onsdigital/zebedee/api/PageTest.java]: I have added the suggested test cases and more. Also refactored the tests to make them clearer 👍

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+    @POST
+    public void createPage(HttpServletRequest request, HttpServletResponse response) {
+
+        // We have to get the request InputStream before reading any request parameters
+        // otherwise the call to get a request parameter will actually consume the body:
+        try (InputStream requestBody = request.getInputStream()) {
+
+            String uri = request.getParameter("uri");
+            if (StringUtils.isEmpty(uri)) {
+                logError(new BadRequestException("uri is empty"));
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: I think you need to add `.log()` to the end

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageCreationHook.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+            logError(e, "failed to create dataset in the dataset api");
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageCreationHook.java]: i link you need to add `.log()`

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageDeletionHook.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+     * @param uri - the URI of the updated page.
+     * @throws IOException
+     */
+    @Override
+    public void onPageUpdated(ApiDatasetLandingPage page, String uri) throws IOException {
+
+        try {
+            datasetAPIClient.deleteDataset(page.getapiDatasetId());
+        } catch (DatasetAPIException e) {
+            logError(e, "failed to delete dataset in the dataset api");
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageDeletionHook.java]: same here

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java (hunk 17 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: Good spot - done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageCreationHook.java (hunk 18 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageCreationHook.java]: done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageDeletionHook.java (hunk 19 above)
Comment (CarlHembrough) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/content/page/APIDatasetLandingPageDeletionHook.java]: done

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+            handleZebdeeException("failed to get collection reader", e, response, uri, session, collection);
+            return;
+        }
+
+        com.github.onsdigital.zebedee.content.page.base.Page page;
+
+        try {
+            page = collectionReader.getContent(uri);
+        } catch (NotFoundException ex) {
+            return; // idempotent
```
Comment (Matt-Guest) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/Page.java]: Can you set the status code here?
